src/mdns/mdns.hpp

```diff
@@ -665,6 +665,59 @@ class Publisher : private NonCopyable
     MdnsTelemetryInfo mTelemetryInfo{};
 };
 
+/**
+ * This interface is a mDNS State Observer.
+ */
+class StateObserver
+{
+public:
+    /**
+     * This method notifies the mDNS state to the observer.
+     *
+     * @param[in] aState  The mDNS State.
+     */
+    virtual void HandleMdnsState(Publisher::State aState) = 0;
+
+    /**
+     * The destructor.
+     */
+    virtual ~StateObserver(void) = default;
+};
+
+/**
+ * This class defines a mDNS State Subject.
+ */
+class StateSubject
+{
+public:
+    /**
+     * Constructor.
+     */
+    StateSubject(void) = default;
+
+    /**
+     * Destructor.
+     */
+    ~StateSubject(void) = default;
+
+    /**
+     * This method adds an mDNS State Observer to this subject.
+     *
+     * @param[in] aObserver  A pointer to the observer. If it's nullptr, it won't be added.
+     */
+    void AddObserver(StateObserver *aObserver);
+
+    /**
+     * This method updates the mDNS State.
+     *
+     * @param[in] aState  The mDNS State.
+     */
+    void UpdateState(Publisher::State aState);
+
+private:
+    std::vector<StateObserver *> mObservers;
```

Comment:
I feel it kind of dangerous to store all the observers using raw pointers. Is it guaranteed that the observers and the subjects all have the same lifetime?

Reply:
Good point! It's guaranteed because all the observers are components of `Application` and they will be released at the end of the program (released by unique_ptr in Application's destructor). Before that the mDNSPublisher has stopped working. I also added `StateSubject::Deinit` to remove all observers in de-initialization which is ahead of the destructors. And in otbr-agent, we won't have any cases of run-time obserser allocation and release.